File: rest/retrofit/src/main/java/at/stefangeyer/challonge/rest/retrofit/RetrofitTournamentRestClient.java
package at.stefangeyer.challonge.rest.retrofit;

import at.stefangeyer.challonge.async.Callback;
import at.stefangeyer.challonge.exception.DataAccessException;
import at.stefangeyer.challonge.model.enumeration.TournamentType;
import at.stefangeyer.challonge.model.query.enumeration.TournamentQueryState;
import at.stefangeyer.challonge.model.query.wrapper.TournamentQueryWrapper;
import at.stefangeyer.challonge.model.wrapper.TournamentWrapper;
import at.stefangeyer.challonge.rest.TournamentRestClient;
import retrofit2.Response;

import java.io.IOException;
import java.time.OffsetDateTime;
import java.util.List;

import static at.stefangeyer.challonge.rest.retrofit.util.RetrofitUtil.callback;
import static at.stefangeyer.challonge.rest.retrofit.util.RetrofitUtil.parse;

/**
 * Retrofit Gson of the tournament rest client
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
public class RetrofitTournamentRestClient implements TournamentRestClient {

    private ChallongeRetrofit challongeRetrofit;

    public RetrofitTournamentRestClient(ChallongeRetrofit challongeRetrofit) {
        this.challongeRetrofit = challongeRetrofit;
    }

    @Override
    public List<TournamentWrapper> getTournaments(TournamentQueryState state, TournamentType type, OffsetDateTime createdAfter,
                                                  OffsetDateTime createdBefore, String subdomain) throws DataAccessException {
        Response<List<TournamentWrapper>> response;

        try {
            response = this.challongeRetrofit.getTournaments(state, type, createdAfter, createdBefore, subdomain).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while getting tournaments", e);
        }

        return parse(response);
    }

    @Override
    public void getTournaments(TournamentQueryState state, TournamentType type, OffsetDateTime createdAfter,
                               OffsetDateTime createdBefore, String subdomain, Callback<List<TournamentWrapper>> onSuccess,
                               Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.getTournaments(state, type, createdAfter, createdBefore, subdomain)
                .enqueue(callback(onSuccess, onFailure, "Error while getting tournaments"));
    }

    @Override
    public TournamentWrapper getTournament(String tournament, boolean includeParticipants,
                                           boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.getTournament(tournament,
                    includeParticipants ? 1 : 0, includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while getting tournament " + tournament, e);
        }

        return parse(response);
    }

    @Override
    public void getTournament(String tournament, boolean includeParticipants, boolean includeMatches,
                              Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.getTournament(tournament, includeParticipants ? 1 : 0, includeMatches ? 1 : 0)
                .enqueue(callback(onSuccess, onFailure, "Error while getting tournament"));
    }

    @Override
    public TournamentWrapper createTournament(TournamentQueryWrapper tournamentData) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.createTournament(tournamentData).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while creating tournament", e);
        }

        return parse(response);
    }

    @Override
    public void createTournament(TournamentQueryWrapper tournamentData, Callback<TournamentWrapper> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.createTournament(tournamentData)
                .enqueue(callback(onSuccess, onFailure, "Error while creating tournament"));
    }

    @Override
    public TournamentWrapper updateTournament(String tournament,
                                              TournamentQueryWrapper tournamentData) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.updateTournament(tournament, tournamentData).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while updating tournament", e);
        }

        return parse(response);
    }

    @Override
    public void updateTournament(String tournament, TournamentQueryWrapper tournamentData,
                                 Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.updateTournament(tournament, tournamentData)
                .enqueue(callback(onSuccess, onFailure, "Error while updating tournament"));
    }

    @Override
    public TournamentWrapper deleteTournament(String tournament) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.deleteTournament(tournament).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while deleting tournament", e);
        }

        return parse(response);
    }

    @Override
    public void deleteTournament(String tournament, Callback<TournamentWrapper> onSuccess,
                                 Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.deleteTournament(tournament)
                .enqueue(callback(onSuccess, onFailure, "Error while deleting tournament"));
    }

    @Override
    public TournamentWrapper processCheckIns(String tournament, boolean includeParticipants,
                                             boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.processCheckIns(tournament,
                    includeParticipants ? 1 : 0, includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while check in processing", e);
        }

        return parse(response);
    }

    @Override
    public void processCheckIns(String tournament, boolean includeParticipants, boolean includeMatches,
                                Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.processCheckIns(tournament,
                includeParticipants ? 1 : 0, includeMatches ? 1 : 0)
                .enqueue(callback(onSuccess, onFailure, "Error while check in processing"));
    }

    @Override
    public TournamentWrapper abortCheckIn(String tournament, boolean includeParticipants,
                                          boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.abortCheckIn(tournament, includeParticipants ? 1 : 0,
                    includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while check in abort", e);
        }

        return parse(response);
    }

    @Override
    public void abortCheckIn(String tournament, boolean includeParticipants, boolean includeMatches,
                             Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.abortCheckIn(tournament, includeParticipants ? 1 : 0, includeMatches ? 1 : 0)
                .enqueue(callback(onSuccess, onFailure, "Error while check in abort"));
    }

    @Override
    public TournamentWrapper startTournament(String tournament, boolean includeParticipants,
                                             boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.startTournament(tournament, includeParticipants ? 1 : 0,
                    includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while starting tournament", e);
        }

        return parse(response);
    }

    @Override
    public void startTournament(String tournament, boolean includeParticipants, boolean includeMatches,
                                Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.startTournament(tournament, includeParticipants ? 1 : 0,
                includeMatches ? 1 : 0).enqueue(callback(onSuccess, onFailure, "Error while starting tournament"));
    }

    @Override
    public TournamentWrapper finalizeTournament(String tournament, boolean includeParticipants,
                                                boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.finalizeTournament(tournament, includeParticipants ? 1 : 0,
                    includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while finalizing tournament", e);
        }

        return parse(response);
    }

    @Override
    public void finalizeTournament(String tournament, boolean includeParticipants, boolean includeMatches,
                                   Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.finalizeTournament(tournament, includeParticipants ? 1 : 0,
                includeMatches ? 1 : 0).enqueue(callback(onSuccess, onFailure, "Error while finalizing tournament"));
    }

    @Override
    public TournamentWrapper resetTournament(String tournament, boolean includeParticipants,
                                             boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.resetTournament(tournament, includeParticipants ? 1 : 0,
                    includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while resetting tournament", e);
        }

        return parse(response);
    }

    @Override
    public void resetTournament(String tournament, boolean includeParticipants, boolean includeMatches,
                                Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.resetTournament(tournament, includeParticipants ? 1 : 0, includeMatches ? 1 : 0)
                .enqueue(callback(onSuccess, onFailure, "Error while resetting tournament"));
    }

    @Override
    public TournamentWrapper openTournamentForPredictions(String tournament, boolean includeParticipants,
                                                          boolean includeMatches) throws DataAccessException {
        Response<TournamentWrapper> response;

        try {
            response = this.challongeRetrofit.openTournamentForPredictions(tournament, includeParticipants ? 1 : 0,
                    includeMatches ? 1 : 0).execute();
        } catch (IOException e) {
            throw new DataAccessException("Error while opening tournament for predictions", e);
        }

        return parse(response);
    }

    @Override
    public void openTournamentForPredictions(String tournament, boolean includeParticipants, boolean includeMatches,
                                             Callback<TournamentWrapper> onSuccess, Callback<DataAccessException> onFailure) {
        this.challongeRetrofit.openTournamentForPredictions(tournament, includeParticipants ? 1 : 0, includeMatches ? 1 : 0)
                .enqueue(callback(onSuccess, onFailure, "Error while opening tournament for predictions"));
    }
}

File: rest/retrofit/src/main/kotlin/at/stefangeyer/challonge/rest/retrofit/RetrofitTournamentRestClient.kt
package at.stefangeyer.challonge.rest.retrofit

import at.stefangeyer.challonge.async.Callback
import at.stefangeyer.challonge.exception.DataAccessException
import at.stefangeyer.challonge.model.enum.TournamentType
import at.stefangeyer.challonge.model.query.enum.TournamentQueryState
import at.stefangeyer.challonge.model.query.wrapper.TournamentQueryWrapper
import at.stefangeyer.challonge.model.wrapper.TournamentWrapper
import at.stefangeyer.challonge.rest.TournamentRestClient
import at.stefangeyer.challonge.rest.retrofit.util.parse
import retrofit2.Call
import retrofit2.Response
import java.time.OffsetDateTime

/**
 * Retrofit gson of the tournament rest client
 *
 * @author Stefan Geyer
 * @version 2018-06-30
 */
class RetrofitTournamentRestClient(private val challongeRetrofit: ChallongeRetrofit) : TournamentRestClient {

    override fun getTournaments(state: TournamentQueryState?, type: TournamentType?, createdAfter: OffsetDateTime?,
                                createdBefore: OffsetDateTime?, subdomain: String?): List<TournamentWrapper> {
        val response = this.challongeRetrofit.getTournaments(state, type, createdAfter, createdBefore, subdomain).execute()
        return response.parse("GetTournaments")
    }

    override fun getTournaments(state: TournamentQueryState?, type: TournamentType?, createdAfter: OffsetDateTime?, createdBefore: OffsetDateTime?, subdomain: String?,
                                onSuccess: Callback<List<TournamentWrapper>>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.getTournaments(state, type, createdAfter, createdBefore, subdomain).enqueue(
                object : retrofit2.Callback<List<TournamentWrapper>> {
                    override fun onFailure(call: Call<List<TournamentWrapper>>, t: Throwable) {
                        onFailure.accept(DataAccessException("GetTournaments request was not successful", t))
                    }

                    override fun onResponse(call: Call<List<TournamentWrapper>>, response: Response<List<TournamentWrapper>>) {
                        onSuccess.accept(response.parse("GetTournaments"))
                    }
                })
    }

    override fun getTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .getTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("GetTournament")
    }

    override fun getTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                               onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.getTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("GetTournament request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("GetTournament"))
                    }
                })
    }

    override fun createTournament(tournamentData: TournamentQueryWrapper): TournamentWrapper {
        val response = this.challongeRetrofit.createTournament(tournamentData).execute()
        return response.parse("CreateTournament")
    }

    override fun createTournament(tournamentData: TournamentQueryWrapper, onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.createTournament(tournamentData).enqueue(object : retrofit2.Callback<TournamentWrapper> {
            override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                onFailure.accept(DataAccessException("CreateTournament request was not successful", t))
            }

            override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                onSuccess.accept(response.parse("CreateTournament"))
            }
        })
    }

    override fun updateTournament(tournament: String, tournamentData: TournamentQueryWrapper): TournamentWrapper {
        val response = this.challongeRetrofit.updateTournament(tournament, tournamentData).execute()
        return response.parse("UpdateTournament")
    }

    override fun updateTournament(tournament: String, tournamentData: TournamentQueryWrapper,
                                  onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.updateTournament(tournament, tournamentData).enqueue(object : retrofit2.Callback<TournamentWrapper> {
            override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                onFailure.accept(DataAccessException("UpdateTournament request was not successful", t))
            }

            override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                onSuccess.accept(response.parse("UpdateTournament"))
            }
        })
    }

    override fun deleteTournament(tournament: String): TournamentWrapper {
        val response = this.challongeRetrofit.deleteTournament(tournament).execute()
        return response.parse("DeleteTournament")
    }

    override fun deleteTournament(tournament: String, onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.deleteTournament(tournament).enqueue(object : retrofit2.Callback<TournamentWrapper> {
            override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                onFailure.accept(DataAccessException("DeleteTournament request was not successful", t))
            }

            override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                onSuccess.accept(response.parse("DeleteTournament"))
            }
        })
    }

    override fun processCheckIns(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .processCheckIns(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("ProcessCheckIns")
    }

    override fun processCheckIns(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                                 onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.processCheckIns(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("ProcessCheckIns request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("ProcessCheckIns"))
                    }
                })
    }

    override fun abortCheckIn(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .abortCheckIn(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("AbortCheckIns")
    }

    override fun abortCheckIn(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                              onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.abortCheckIn(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("AbortCheckIns request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("AbortCheckIns"))
                    }
                })
    }

    override fun startTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .startTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("StartTournament")
    }

    override fun startTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                                 onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.startTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("StartTournament request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("StartTournament"))
                    }
                })
    }

    override fun finalizeTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .finalizeTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("FinalizeTournament")
    }

    override fun finalizeTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                                    onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.finalizeTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("FinalizeTournament request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("FinalizeTournament"))
                    }
                })
    }

    override fun resetTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .resetTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("ResetTournament")
    }

    override fun resetTournament(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                                 onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.finalizeTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("ResetTournament request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("ResetTournament"))
                    }
                })
    }

    override fun openTournamentForPredictions(tournament: String, includeParticipants: Boolean, includeMatches: Boolean): TournamentWrapper {
        val response = this.challongeRetrofit
                .openTournamentForPredictions(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .execute()
        return response.parse("OpenTournamentForPredictions")
    }

    override fun openTournamentForPredictions(tournament: String, includeParticipants: Boolean, includeMatches: Boolean,
                                              onSuccess: Callback<TournamentWrapper>, onFailure: Callback<DataAccessException>) {
        this.challongeRetrofit.finalizeTournament(tournament, if (includeParticipants) 1 else 0, if (includeMatches) 1 else 0)
                .enqueue(object : retrofit2.Callback<TournamentWrapper> {
                    override fun onFailure(call: Call<TournamentWrapper>, t: Throwable) {
                        onFailure.accept(DataAccessException("OpenTournamentForPredictions request was not successful", t))
                    }

                    override fun onResponse(call: Call<TournamentWrapper>, response: Response<TournamentWrapper>) {
                        onSuccess.accept(response.parse("OpenTournamentForPredictions"))
                    }
                })
    }
}